File: rexe-backend/containers/cpp/stub.cpp
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <stdlib.h>
#include <iostream>
#include <fstream>
#include <assert.h>

class rexeCppStub {
public:
  void setlimits() {
    rlimit limit;
    long bytes = (1024 * 1024) * %MEMORY_LIMIT%;
    long time = %TIME_LIMIT%;

    limit.rlim_cur = bytes;
    limit.rlim_max = bytes;

    if (setrlimit(RLIMIT_AS, &limit) != 0) {
      printf("Error: %s\n", strerror(errno));
      exit(1);    
    }

    if (getrlimit(RLIMIT_AS, &limit) != 0) {
      printf("Error: %s\n", strerror(errno));
      exit(1);
    }

    assert(limit.rlim_cur == bytes && limit.rlim_max == bytes);

    limit.rlim_cur = time;
    limit.rlim_max = time;

    if (setrlimit(RLIMIT_CPU, &limit) != 0) {
      printf("Error: %s\n", strerror(errno));
      exit(1);    
    }

    if (getrlimit(RLIMIT_CPU, &limit) != 0) {
      printf("Error: %s\n", strerror(errno));
      exit(1);
    }

    assert(limit.rlim_cur == time && limit.rlim_max == time);
  }
  void writeResourceUsage() {
    rusage usage;
    if (getrusage(RUSAGE_SELF, &usage) != 0) {
      printf("Error: %s\n", strerror(errno));
      exit(1);
    }
    std::ofstream file("rusage.txt");
    file << (usage.ru_utime.tv_usec / 1000.0) << "," << (usage.ru_maxrss / 1024.0);
    file.close();    
  }    
  rexeCppStub() {
    setlimits();
  }
  ~rexeCppStub() {
    writeResourceUsage();
  }
}_rexeCppStub;

File: rexe-backend/containers/py/stub.py
from functools import wraps
import resource

class rexePyStub:
  def setlimits(self):
    bytes = (1024 * 1024) * %MEMORY_LIMIT%
    time = %TIME_LIMIT%
    rsrcs = [resource.RLIMIT_AS, resource.RLIMIT_DATA, resource.RLIMIT_STACK]
    for rsrc in rsrcs:
      resource.setrlimit(rsrc, (bytes, bytes))
      soft, hard = resource.getrlimit(rsrc)
      assert soft == bytes and hard == bytes

    resource.setrlimit(resource.RLIMIT_CPU, (time, time))
    soft, hard = resource.getrlimit(resource.RLIMIT_CPU)
    assert soft == time and hard == time

  def writeResourceUsage(self):
    rusage = resource.getrusage(resource.RUSAGE_SELF)
    f = open('rusage.txt', 'w')
    f.write(str(rusage[0] * 1000.0) + ',' + str(rusage[2] / 1024.0))
    f.close()

  def __init__(self):
    self.setlimits()

  def __del__(self):
    self.writeResourceUsage()

rexePyStub() 
